script/model/training_util/data_transforming_util.py

```diff
@@ -44,63 +45,98 @@ def _tuple_num_memory_cardinality_linear_predict_transform(x, y):
-    return y * (tuple_num * np.log2(tuple_num) + 1)[:, np.newaxis]
+    return y * (tuple_num * np.log2(tuple_num) + _TRANSFORM_EPSILON)[:, np.newaxis]
 
 
 # Transform the target in a linear-log way (nlogn) according to the tuple num
-_tuple_num_linear_log_transformer = (_tuple_num_linear_log_train_transform, _tuple_num_linear_log_predict_transform)
+_num_linear_log_transformer = (_num_linear_log_train_transform, _num_linear_log_predict_transform)
 
 
-def _tuple_num_log_train_transform(x, y):
+def _num_log_train_transform(x, y):
     # Transform down the target according to the log tuple num value in the input
     tuple_num = np.copy(x[:, data_info.INPUT_CSV_INDEX[ExecutionFeature.NUM_ROWS]])
-    return y / (np.log2(tuple_num) + 1)[:, np.newaxis]
+    return y / (np.log2(tuple_num) + _TRANSFORM_EPSILON)[:, np.newaxis]
 
 
-def _tuple_num_log_predict_transform(x, y):
+def _num_log_predict_transform(x, y):
     # Transform up the target according to the log tuple num value in the input
     tuple_num = np.copy(x[:, data_info.INPUT_CSV_INDEX[ExecutionFeature.NUM_ROWS]])
-    return y * (np.log2(tuple_num) + 1)[:, np.newaxis]
+    return y * (np.log2(tuple_num) + _TRANSFORM_EPSILON)[:, np.newaxis]
 
 
 # Transform the target in a log scale (logn) according to the tuple num
-_tuple_num_log_transformer = (_tuple_num_log_train_transform, _tuple_num_log_predict_transform)
+_num_log_transformer = (_num_log_train_transform, _num_log_predict_transform)
+
+
+def _num_cardinality_linear_log_train_transform(x, y):
+    # Transform down the target according to the cardinality * log tuple num value in the input
+    tuple_num = np.copy(x[:, data_info.INPUT_CSV_INDEX[ExecutionFeature.NUM_ROWS]])
+    cardinality = np.copy(x[:, data_info.INPUT_CSV_INDEX[ExecutionFeature.EST_CARDINALITIES]])
+    return y / ((cardinality * np.log2(tuple_num)) + _TRANSFORM_EPSILON)[:, np.newaxis]
+
+
+def _num_cardinality_linear_log_predict_transform(x, y):
+    # Transform up the target according to the cardinality * log tuple num value in the input
+    tuple_num = np.copy(x[:, data_info.INPUT_CSV_INDEX[ExecutionFeature.NUM_ROWS]])
+    cardinality = np.copy(x[:, data_info.INPUT_CSV_INDEX[ExecutionFeature.EST_CARDINALITIES]])
+    return y * ((cardinality * np.log2(tuple_num)) + _TRANSFORM_EPSILON)[:, np.newaxis]
+
+
+# Transform the target in a linear log scale (cardinality * log tuple_num)
+_num_cardinality_linear_log_transformer = (_num_cardinality_linear_log_train_transform,
+                                           _num_cardinality_linear_log_predict_transform)
+
+def _cardinality_linear_train_transform(x, y):
+    # Transform down the target according to the cardinality in the input
+    cardinality = np.copy(x[:, data_info.INPUT_CSV_INDEX[ExecutionFeature.EST_CARDINALITIES]])
+    return y / (cardinality + _TRANSFORM_EPSILON)[:, np.newaxis]
+
+
+def _cardinality_linear_predict_transform(x, y):
```

Comment:
Eh sorry that I was not clear before. I didn't mean to remove the word "tuple" from all functions. I was only referring to this function, since this only transforms with the cardinality. The functions that transform with tuple num should still has "tuple_num" in their names... Or even better, we can name "tuple_num" to "num_rows" to be consistent with the csv header.